File: server/src/main/java/org/dan/jadalnia/app/festival/FestivalDao.java
package org.dan.jadalnia.app.festival;

import com.google.common.collect.ImmutableMap;
import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;
import lombok.experimental.FieldDefaults;
import lombok.extern.slf4j.Slf4j;
import org.dan.jadalnia.app.festival.menu.MenuItem;
import org.dan.jadalnia.app.festival.pojo.FestivalInfo;
import org.dan.jadalnia.app.festival.pojo.FestivalState;
import org.dan.jadalnia.app.festival.pojo.Fid;
import org.dan.jadalnia.app.festival.pojo.NewFestival;
import org.dan.jadalnia.sys.ctx.FutureExecutor;
import org.jooq.DSLContext;

import javax.inject.Inject;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.CompletableFuture;

import static java.util.Optional.ofNullable;
import static org.dan.jadalnia.jooq.Tables.FESTIVAL;
import static org.dan.jadalnia.sys.error.JadEx.internalError;
import static org.dan.jadalnia.sys.error.JadEx.notFound;

@Slf4j
@RequiredArgsConstructor(onConstructor = @__(@Inject))
@FieldDefaults(level = AccessLevel.PRIVATE, makeFinal = true)
public class FestivalDao {
    DSLContext jooq;
    FutureExecutor executor;

    public CompletableFuture<Fid> create(NewFestival newTournament) {
        return executor.run(() -> jooq.insertInto(FESTIVAL,
                FESTIVAL.STATE,
                FESTIVAL.OPENS_AT,
                FESTIVAL.MENU,
                FESTIVAL.NAME)
                .values(FestivalState.Announce,
                        newTournament.getOpensAt(),
                        Collections.emptyList(),
                        newTournament.getName())
                .returning(FESTIVAL.FID)
                .fetchOne()
                .getFid());
    }

    public CompletableFuture<Void> setState(Fid fid, FestivalState state) {
        return jooq.update(FESTIVAL)
                .set(FESTIVAL.STATE, state)
                .where(FESTIVAL.FID.eq(fid))
                .executeAsync(executor.getExecutorService())
                .thenAccept(n -> {
                    if (n == 0) {
                        throw internalError(
                                "state :state is not set for festival :fid",
                                ImmutableMap.of("state", state, "fid", fid));
                    }
                })
                .toCompletableFuture();
    }

    public CompletableFuture<Integer> setMenu(Fid fid, List<MenuItem> items) {
        return executor.run(
                () -> jooq.update(FESTIVAL)
                        .set(FESTIVAL.MENU, items)
                        .where(FESTIVAL.FID.eq(fid))
                        .execute())
                .toCompletableFuture();
    }

    public CompletableFuture<FestivalInfo> getById(Fid fid) {
        return executor.run(() ->
                ofNullable(jooq.select()
                        .from(FESTIVAL)
                        .where(FESTIVAL.FID.eq(fid))
                        .fetchOne())
                        .map(r -> new FestivalInfo(
                                        fid,
                                        r.get(FESTIVAL.NAME),
                                        r.get(FESTIVAL.STATE),
                                        r.get(FESTIVAL.MENU),
                                        r.get(FESTIVAL.OPENS_AT)))
                        .orElseThrow(() -> notFound("Festival not found" , "fid", fid)));
    }
}

File: service/src/main/kotlin/org/dan/jadalnia/app/festival/FestivalDao.kt
package org.dan.jadalnia.app.festival

import com.google.common.collect.ImmutableMap
import org.dan.jadalnia.app.festival.menu.MenuItem
import org.dan.jadalnia.app.festival.pojo.FestParams
import org.dan.jadalnia.app.festival.pojo.FestivalInfo
import org.dan.jadalnia.app.festival.pojo.FestivalState
import org.dan.jadalnia.app.festival.pojo.Fid
import org.dan.jadalnia.app.festival.pojo.NewFestival
import org.dan.jadalnia.app.label.AsyncDao
import org.dan.jadalnia.jooq.Tables.FESTIVAL
import org.dan.jadalnia.sys.error.JadEx.Companion.internalError
import org.dan.jadalnia.sys.error.JadEx.Companion.notFound
import java.util.Collections

import java.util.Optional.ofNullable
import java.util.concurrent.CompletableFuture

class FestivalDao : AsyncDao() {
  fun create(newFestival: NewFestival): CompletableFuture<Fid> {
    return execQuery { jooq ->
      jooq.insertInto(FESTIVAL,
          FESTIVAL.STATE,
          FESTIVAL.OPENS_AT,
          FESTIVAL.MENU,
          FESTIVAL.NAME)
          .values(FestivalState.Announce,
              newFestival.opensAt,
              Collections.emptyList(),
              newFestival.name)
          .returning(FESTIVAL.FID)
          .fetchOne()
          .getFid()
    }
  }

  fun setState(fid: Fid, state: FestivalState): CompletableFuture<Void> {
    return execQuery { jooq ->
      jooq.update(FESTIVAL)
          .set(FESTIVAL.STATE, state)
          .where(FESTIVAL.FID.eq(fid))
          .execute()
    }
        .thenAccept { n ->
          if (n == 0) {
            throw internalError("state :state is not set for festival :fid",
                ImmutableMap.of("state", state, "fid", fid));
          }
        }
  }

  fun setMenu(fid: Fid, items: List<MenuItem>): CompletableFuture<Int> {
    return execQuery { jooq ->
      jooq.update(FESTIVAL)
          .set(FESTIVAL.MENU, items)
          .where(FESTIVAL.FID.eq(fid))
          .execute()
    }
  }

  fun getById(fid: Fid): CompletableFuture<FestivalInfo> {
    return execQuery { jooq ->
      ofNullable(jooq.select()
          .from(FESTIVAL)
          .where(FESTIVAL.FID.eq(fid))
          .fetchOne())
          .map { r ->
            FestivalInfo(
                fid = fid,
                name = r.get(FESTIVAL.NAME),
                state = r.get(FESTIVAL.STATE),
                menu = r.get(FESTIVAL.MENU),
                params = r.get(FESTIVAL.PARAMS) ?: FestParams(),
                opensAt = r.get(FESTIVAL.OPENS_AT))
          }
          .orElseThrow { notFound("Festival not found", "fid", fid) }
    }
  }

  fun updateParams(fid: Fid, params: FestParams)
      = execQuery { jooq ->
    jooq.update(FESTIVAL)
        .set(FESTIVAL.PARAMS, params)
        .where(FESTIVAL.FID.eq(fid))
        .execute()
  }
}
